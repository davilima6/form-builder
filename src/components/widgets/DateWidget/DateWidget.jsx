// @flow
import * as React from 'react';
import styled, { css } from 'styled-components';
import { normalize } from '../../../utils';

const { useCallback } = React;

const WidgetWrapper = styled.div`
  ${props => props.error
    && css`
      border-color: var(--dark-red);
    `};
`;

const Label = styled.label``;

const StyledLabel = styled.div`
  flex: 0 0 100%;
  font-weight: 500;
  margin-bottom: 0.5em;
`;

const StyledInput = styled.input`
  font-size: 1em;
  padding: 3px 5px;
  width: 100%;
`;

const StyledError = styled.div`
  border: 1px dashed red;
`;

type Props = {
  id: string,
  label: string,
  required?: boolean,
  value?: number,
  errors?: Array<string>,
  onChangeField: Function,
};

const DateWidget = ({
  id, label, required, value, errors, onChangeField,
}: Props) => {
  const fieldId = `field-${id}`;

  const onChange = useCallback(({ target }: SyntheticEvent<>) => onChangeField(id, target.value), [
    id,
    onChangeField,
  ]);

  return (
    <WidgetWrapper error={errors.length}>
      <Label htmlFor={fieldId}>
        <StyledLabel>{label}</StyledLabel>
        <StyledInput
          type="date"
          id={fieldId}
          name={id}
          required={required}
          value={value}
          onChange={onChange}
        />
      </Label>
      {errors.map(message => (
        <StyledError key={normalize(message)}>{message}</StyledError>
      ))}
    </WidgetWrapper>
  );
};

DateWidget.defaultProps = {
  required: false,
  value: '',
  errors: [],
};

export default DateWidget;
