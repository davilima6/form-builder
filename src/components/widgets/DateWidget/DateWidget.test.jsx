import React from 'react';
import { cleanup, render } from 'react-testing-library';
import 'jest-dom/extend-expect';
// import 'jest-styled-components';
import DateWidget from './DateWidget';

describe('DateWidget', () => {
  let component;

  beforeEach(() => {
    const props = {
      label: 'Date of Birth',
      type: 'text',
      data: new Date(0).getTime(),
      errors: [],
      onChange: () => {},
    };

    cleanup();
    component = render(<DateWidget {...props} />);
  });

  test('renders without crashing', () => expect(component).toBeDefined());

  test.skip('matches snapshot', () => {
    const { asFragment } = component;

    expect(asFragment()).toMatchSnapshot();
    expect(asFragment()).toHaveStyleRule('background', '#fff');
  });
});
