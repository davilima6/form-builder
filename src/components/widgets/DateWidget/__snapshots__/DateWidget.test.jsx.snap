// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateWidget matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa djLaIN"
  >
    <label
      class="sc-bwzfXH gezcZz"
      for="field-undefined"
    >
      <span
        class="label"
      >
        Date of Birth
      </span>
      <input
        id="field-undefined"
        type="datetime-local"
        value=""
      />
    </label>
  </div>
</DocumentFragment>
`;
